openpgp/packet/private_key.go

```diff
@@ -258,6 +258,11 @@ func (pk *PrivateKey) Serialize(w io.Writer) (err error) {
 	if err != nil {
 		return
 	}
+
+	if !pk.Encrypted && !pk.Dummy() {
+		pk.s2kType = S2KNON
+	}
+
```

Comment:
Since `S2KNON == 0`, and `0` is the default value of `pk.s2kType`, the only possible way this can happen is after a `Decrypt`, right? So maybe the proper way to fix this is to set `pk.s2kType = S2KNON` in `PrivateKey.Decrypt()`?

Reply:
I am not sure, some GopenPGP tests fail if we put `pk.s2kType = S2KNON` in `PrivateKey.Decrypt()`

Reply:
Because we duplicate via serialize that is not decryption

Reply:
@wussler Yeah but, in `Serialize`, the `s2kType` already has to come from somewhere, if it is not `S2KNON` (since that is the default). I can only see it coming from `parse` (in which case it's encrypted if it's not `S2KNON`) or `Encrypt`. So basically `!pk.Encrypted && !pk.Dummy() && pk.s2kType != S2KNON` should never happen in my opinion, it violates invariants. The only way I could imagine it happening is if we first parse the key, then decrypt, then serialize, but I might be missing something.

@zugzwang I guess could you try to find out where the `s2kType` comes from? :grimacing:

Reply:
@twiss Yeah I agree, I think it should come from `parse`. I was just referring that it is indeed working correctly on key decryption, and rather failing on key duplication. But it could be an error carried over from decryption too :crying_cat_face:

Reply:
Let me point out that this is what the code was doing before the PR, I just reproduced it as a hot fix for the serialization issue.

I had a look, `s2kType` is unexported and only used in private_key.go. 
When calling `Encrypt`, it gets set to `S2KSHA1` or `S2KCHECKSUM`. This is the only place where it is set.

But then, `PrivateKey.Encrypted` is exported, so a client is free to set `Encrypted` to whatever, and then we get this behavior. On x/crypto 's tests, keys look consistent \*, but in GopenPGP tests, indeed we get keys `!Encrypted && s2kType != S2KNON`.

Conversely, `x/crypto` is setting `Encrypted = false` only when successfully parsing an unencrypted key or when parsing a key with `s2kType == S2KNON` so it seems right.

\* Minus a potential issue with dummy keys I have just spotted

Reply:
OK, LGTM. It is also a good reason that you set the s2kType when encrypting, since you might want to change it, and not reuse the value from decryption. 
@twiss any opinion?